Prohibit attribute value from evolving to contain complex types (#3858)
open-telemetry/opentelemetry-specification: specification/common/README.md

both containing an array of strings to represent a mapping
`header_keys[i] -> header_values[i]`).

> Attribute values MUST NOT be evolved to support complex types (e.g. maps and

Reviewer: I think this is confusing because the spec does not usually address spec writers.

```suggestion
Note for spec authors: the OpenTelemetry project has decided to never support complex types (e.g. maps and
```

Author: Went in a different direction after PR feedback and this recommendation is no longer relevant.

> Attribute values MUST NOT be evolved to support complex types (e.g. maps and

Reviewer 2: I understand what we are trying to say here, but I think this wording is overly broad and might be better stated that the value `type` defined by the specification for attributes will only be defined for base primitive types and will not support complex or nested values.

And explicitly call out that for systems / applications / exporters that may wish to represent or send complex types that an encoding scheme may be defined to support a common encoding / decoding (to support cross system compatibility). So a SDK (or user defined (not semantic convention) attribute value) can be encoded and sent that could be deterministically identified and decoded by any downstream receiver.

So I think we should be explicit about the `type` being of a primitive type, but a little vague about the "how" the encoding may (or may not) be implemented (until we actually do identify / implement). This is so that once a scheme is defined it "may" be possible that SDK implementation(s) or applications will perform this encoding with or without encoding / decoding helpers. eg. they might just be helpers or specific SDK implementations that when a non-primitive is passed it gets automatically encoded into some defined common scheme.

Author: Went in a different direction after PR feedback and this comment is no longer relevant. Please take a look.

> heterogeneous arrays). Note: the log data model

Reviewer 3: nit (editorial):

```suggestion
heterogeneous arrays). Note: the Logs Data Model
```

> defines [attributes](../logs/data-model.md#field-attributes) with values of

Reviewer 3: ```suggestion
defines [Attributes field](../logs/data-model.md#field-attributes) with values of
```

> type `Any`. This is an intentional divergence from the standard definition of

Reviewer 3: ```suggestion
type `any`. This is an intentional divergence from the standard definition of
```

> attributes defined here. Log record attributes are a super set of standard

Reviewer 3: ```suggestion
attributes defined here. Log record attributes are a superset of standard
```

> attributes.

Reviewer 4: It would be good to clarify that this is for compatibility with existing logging systems that may produce complex attributes.

Reviewer 3: This is called out in https://github.com/open-telemetry/opentelemetry-specification/pull/3852

See [this document](attribute-type-mapping.md) to find out how to map values obtained
outside OpenTelemetry into OpenTelemetry attribute values.

**[1]**: Note: extending the set of allowable attribute value types is a

> breaking change.

Author: @tedsuo suggest extending this note with the following text:

> OpenTelemetry limits the definition of standard attributes for two reasons.

> One reason is practical. We have existing API fields which conform to this subset of attribute types. For these fields, backends do not expect additional attribute types. It would be a breaking change to expose additional value types on these fields because we would be breaking a contract we have established with our users.

> The second reason is a design choice. In OpenTelemetry we have made a decision regarding how we define metadata. Any time we need metadata in an API, we create use standard attributes, ideally in a field named "Attributes." OpenTelemetry defines “metadata” as a field explicitly intended to be used to create search indexes and statistical data.

> We have made the design choice that flat, simple data is appropriate for this use case. We do not want end users and backends to have to dig into some kind of complex type in order to create these indexes and statistics. It should be possible for an automated system to make use of metadata without having to make a decision as to how to handle complex types. We believe that simple, flat data best enables these use cases.

> To that end, OpenTelemetry SHOULD NOT define semantic conventions for standard attributes that are essentially complex data type encoded into a simple data types. For example, a string encoded JSON attribute is not an acceptable semantic convention for a standard attribute in OpenTelemetry.

Starting a thread so we can discuss the language.

Author: I would want to say something to the effect of:

- With respect to the issue of whether adding additional attribute value types is a breaking change, there are arguments on both sides. We proposed extending attribute values to support complex types, and received significant pushback from the community. Rather than being lost in the grey area indefinitely, we chose to explicitly define additional value types as a breaking change.
- Limiting the types of attribute values to a set which has proved sufficient during the several years of OpenTelemetry's development is a useful guardrail for design. In taking additional value types off the table, we narrow the solution space and reduce bikeshedding.
- The decision to limit attribute values to primitives and arrays of primitives supports OpenTelemetry's intended interpretation of "attributes are metadata", rather than "attributes are data". An "attributes are data" interpretation suggests unbounded value types but creates problems for consumers of the data which have to be prepared to receive any type of data structure. An "attributes are metadata" interpretation suggests a bounded set of value types and discouragement of encoding complex data in attributes (i.e. string encoded JSON is an abuse of attributes). This facilitates data consumers' ability to create search indexes and perform other statistical analysis.

Reviewer 5: Is this proposal only to take complex types off the table, or also to take other primitive types off the table (e.g. datetime, bytes, ...)?

Author: The current proposal is that any additional to the value types, primitive or otherwise, is considered a breaking change. We could change it to say carve out an exception for adding additional primitive types, but then we have to define what a primitive type is. 

Personally, I'm comfortable being limited to the existing set of primitives. They've gotten us this far and is reasonable to imagine that representing additional types of primitive-like data (e.g. datetimes) can be defined as an encoding on top of one of the existing types (i.e. a datetime can be an ISO 8601 string or 64 bit long of epoch millis). Even bytes could be represented using base64 encoded strings. But I would argue that supporting a bytes value type is in conflict with the "attributes are metadata" interpretation.

Reviewer 6: >"attributes are metadata" interpretation.

Do we have that interpretation defined somewhere in the spec? I don't understand what we call metadata in this context.

Author: > Do we have that interpretation defined somewhere in the spec?

No, not today. But in the 2/20 spec meeting we proposed including an explanation of the decision in the spec, which might include [text like this](https://github.com/open-telemetry/opentelemetry-specification/pull/3858#discussion_r1496574959) explaining the attributes are metadata interpretation:

>The decision to limit attribute values to primitives and arrays of primitives supports OpenTelemetry's intended interpretation of "attributes are metadata", rather than "attributes are data". An "attributes are data" interpretation suggests unbounded value types but creates problems for consumers of the data which have to be prepared to receive any type of data structure. An "attributes are metadata" interpretation suggests a bounded set of value types and discouragement of encoding complex data in attributes (i.e. string encoded JSON is an abuse of attributes). This facilitates data consumers' ability to create search indexes and perform other statistical analysis.

Reviewer 7: I'm also concerned about OTel defining (or redefining) metadata...which dictionary.com says is simply "a set of data that describes and gives information about other data." Can we be more concise on the reason? Reading Ted's paragraph, it seems like it primarily boils down to "making it easier for backends to ingest and index data".

Author: @tigrannajaryan / @mattmccleary - so I understand correctly, do you think we shouldn't mention metadata as part of the explanation of why no complex types? Or do you think its ok to mention metadata, but that we should be brief to minimize chance of confusion?

Reviewer 7: Yes, my suggestion is the first - leave out any explanation of what we consider metadata and make it as brief and focused as possible...

Reviewer 6: >do you think we shouldn't mention metadata as part of the explanation of why no complex types? Or do you think its ok to mention metadata, but that we should be brief to minimize chance of confusion?

I think we shouldn't unless we clearly define what is "metadata" and I don't think I see a clear definition of it.

Reviewer 6: Sorry Jack, I don't think that helps. I actually find it worse, to mention metadata without explaining what it is. The common definition I am aware of is that metadata is "data about other data" and I have no idea how to apply that in this context.

I am confused and really don't understand what is it that we are trying to say here.

Author: How about:

> Limiting attribute value types to primitives and arrays of primitives supports OpenTelemetry's intent that attributes are meant to be used to create search indexes and perform other statistical analysis. They are NOT intended to carry data of unbounded complexity.

Reviewer 7: Is this what we're trying to say (even shorter):

Limiting attribute value types to primitives and arrays of primitives simplifies data consumers' efforts to create search indexes and perform statistical analysis.

Reviewer 6: That is clear, thank you.

### Standard Attribute

Attributes are used in various places throughout the OpenTelemetry data model.

> We designate the [previous attribute section](#attribute) as the standard

Reviewer 7: Why not just update the previous section title to say "Standard Attribute"? Feels weird to "redesignate" previous section ex post facto.

Author: I thought about this but decided not to for a couple reasons:
- The `./README#attribute` heading is linked to in 27 different places. Updating the link text of all these places to be "standard attributes" makes for a worse reading experience. Its nice to be able to have a shorthand.
- I was thinking of this section sort like a glossary entry. It clarifies and some context around the decision, and provides a nice anchor for any spec that wants to specifically refer the difference between standard attributes and some extended version. If its all rolled into one section, you might be inclined to refer to a specific paragraph for this information.

I think the redesignation in this section is more like adding an alias.

Author: Marking this as resolved. Please re-open if you have additional thoughts.

See [this document](attribute-type-mapping.md) to find out how to map values obtained
outside OpenTelemetry into OpenTelemetry attribute values.

> **[1]**: NOTE: extending the set of attribute value types is a breaking change.

Reviewer 8: As this PR has been discussed, I will create an issue to address this clarification, as a follow up.
